src/lib/ams.c

```diff
@@ -0,0 +1,749 @@
+				     uint32_t incoming_slot)
+{
+	bool found_any = false;
+	bool incoming = (incoming_slot != AMS_INVALID_SLOT);
+	struct ams_routing_entry *routing_table;
+	uint32_t forwarded;
+	uint32_t slot;
+	struct ams_routing_entry am_target;
+	uint32_t ixc_route;
+	int err = 0;
+
+	if (ams->am_service_context == NULL || ams_message_payload == NULL)
+		return -EINVAL;
+
+
+	ams_acquire(ams->am_service_context);
+
+	if (incoming)
+		ctx.shared->slot_done[incoming_slot] |= cpu_get_id();
+
+	routing_table = ams->am_service_context->shared->routing_table;
+
+	/* first search for external consumer and try to reserve slot right away */
+	slot = AMS_INVALID_SLOT;
+	for (uint32_t iter = 0; iter < ams->am_service_context->shared->routing_table_size; iter++) {
+		if (routing_table[iter].message_type_id != ams_message_payload->message_type_id)
+			continue;
+
+		/* check if we want to limit to specific module */
+		if (module_id != AMS_ANY && instance_id != AMS_ANY) {
+			if (routing_table[iter].consumer_module_id != module_id ||
+			    routing_table[iter].consumer_instance_id != instance_id) {
+				continue;
+			}
+		}
+		am_target = routing_table[iter];
+		ixc_route = am_service_get_ixc_route_to_target(cpu_get_id(),
+							       am_target.consumer_core_id);
+		if (ixc_route != cpu_get_id()) {
+			if (incoming) {
+				slot = incoming_slot;
+			} else {
+				slot = am_service_push_slot(ams->am_service_context,
+							    ams_message_payload, module_id,
+							    instance_id);
+				if (slot == AMS_INVALID_SLOT) {
+					ams_release(ams->am_service_context);
+					return -EINVAL;
+				}
+			}
+		}
+	}
+
+	for (uint32_t iter = 0; iter < ams->am_service_context->shared->routing_table_size; iter++) {
+		/* Search for required entry */
+		if (routing_table[iter].message_type_id != ams_message_payload->message_type_id)
+			continue;
+
+		/* check if we want to limit to specific module* */
+		if (module_id != AMS_ANY && instance_id != AMS_ANY) {
+			if (routing_table[iter].consumer_module_id != module_id ||
+			    routing_table[iter].consumer_instance_id != instance_id) {
+				continue;
+			}
+		}
+
+		found_any = true;
+		am_target = routing_table[iter];
+		ixc_route = am_service_get_ixc_route_to_target(cpu_get_id(),
+							       am_target.consumer_core_id);
+
+		if (ixc_route == cpu_get_id()) {
+			/* we are on target core alredy */
+			/* release lock here, callback are NOT supposed to change routing_table */
+			ams_release(ams->am_service_context);
+
+			am_target.consumer_callback(ams_message_payload, am_target.ctx);
+			err = 0;
+		} else {
+			/* we have to go thru idc */
+			if (incoming) {
+				/* if bit is set we are forwarding it again */
+				if (ams->am_service_context->shared->slot_done[incoming_slot] & (1 << am_target.consumer_core_id)) {
+					/* slot was already processed for that core, skip it */
+					continue;
+				}
+			}
+			/* if bit cleared */
+			if (forwarded & (~(1 << am_target.consumer_core_id))) {
```

Comment:
the comment and the test don't match. The test checks if any other bit is set, not if a specific bit is cleared